src/core/ext/transport/chaotic_good/server/chaotic_good_server.cc

```diff
@@ -0,0 +1,370 @@
+          args->endpoint),
+      SliceBuffer());
+  // Start receiving setting frames.
+  self->connection_->receive_settings_activity_ = OnReceive(self);
+}
+
+ActivityPtr
+ChaoticGoodServerListener::ActiveConnection::HandshakingState::OnReceive(
+    HandshakingState* self) {
+  auto on_receive_frames = TrySeq(
+      self->connection_->endpoint_->ReadSlice(FrameHeader::frame_header_size_),
+      [self = self](Slice slice) mutable {
+        // Parse frame header
+        auto frame_header = std::make_shared<FrameHeader>(
+            FrameHeader::Parse(reinterpret_cast<const uint8_t*>(
+                                   GRPC_SLICE_START_PTR(slice.c_slice())))
+                .value());
+        return TrySeq(
+            self->connection_->endpoint_->Read(frame_header->GetFrameLength()),
+            [frame_header = frame_header,
+             self = self](SliceBuffer buffer) mutable {
+              // Read Setting frame.
+              SettingsFrame frame;
+              // Initialized to get this_cpu() info in global_stat().
+              ExecCtx exec_ctx;
+              // Deserialize frame from read buffer.
+              absl::BitGen bitgen;
+              auto status = frame.Deserialize(
+                  self->connection_->hpack_parser_.get(), *frame_header,
+                  absl::BitGenRef(bitgen), buffer);
+              GPR_ASSERT(status.ok());
+              self->connection_->connection_type_ =
+                  frame.headers
+                      ->get_pointer(ChaoticGoodConnectionTypeMetadata())
+                      ->Ref();
+              bool is_control_endpoint =
+                  std::string(
+                      self->connection_->connection_type_.as_string_view()) ==
+                  "control";
+              if (!is_control_endpoint) {
+                // Get connection id for data endpoint.
+                self->connection_->connection_id_ =
+                    frame.headers
+                        ->get_pointer(ChaoticGoodConnectionIdMetadata())
+                        ->Ref();
+              }
+              return If(
+                  is_control_endpoint,
+                  // connection_type = control
+                  [self = self]() mutable {
+                    self->connection_->GenerateConnectionID();
+                    // Add a wait latch for data endpoint to connect.
+                    auto latch = std::make_shared<
+                        Latch<std::shared_ptr<PromiseEndpoint>>>();
+                    {
+                      MutexLock lock(&self->connection_->listener_->mu_);
+                      self->connection_->listener_->connectivity_map_.insert(
+                          std::pair<std::string,
+                                    std::shared_ptr<Latch<
+                                        std::shared_ptr<PromiseEndpoint>>>>(
+                              self->connection_->connection_id_
```

Comment:
Since we anyway need to check for collisions, recommend changing `GenerateConnectionID` to be `NewConnectionID` and doing the insertion in that function too (all under the same mutex - including the bitgen)